src/tablenavigation.js

```diff
@@ -0,0 +1,515 @@
+
+		if ( objectElement && model.schema.isObject( objectElement ) ) {
+			return;
+		}
+
+		// If next to the selection there is an object then this is not the cell boundary (widget handler should handle this).
+		if ( this._getObjectElementNextToSelection( selection, isForward ) ) {
+			return;
+		}
+
+		// If there isn't any $text position between cell edge and selection then we shall move the selection to next cell.
+		const textRange = this._findTextRangeFromSelection( cellRange, selection, isForward );
+
+		if ( !textRange ) {
+			this._navigateFromCellInDirection( tableCell, direction );
+
+			return true;
+		}
+
+		// If the navigation is horizontal then we have no more custom cases.
+		if ( [ 'left', 'right' ].includes( direction ) ) {
+			return;
+		}
+
+		// If the range is a single line then move the selection to the beginning/end of a cell content.
+		//
+		// We can't move the selection directly to the another cell because of dual position at the end/beginning
+		// of wrapped line (it's at the same time at the end of one line and at the start of the next line).
+		if ( this._isSingleLineRange( textRange, isForward ) ) {
+			model.change( writer => {
+				writer.setSelection( isForward ? cellRange.end : cellRange.start );
+			} );
+
+			return true;
+		}
+	}
+
+	/**
+	 * Returns `true` if `selection` is at `cellRange` edge according to navigation `direction`.
+	 *
+	 * @private
+	 * @param {module:engine/model/range~Range} cellRange The bounding cell range.
+	 * @param {module:engine/model/selection~Selection} selection The current selection.
+	 * @param {Boolean} isForward The expected navigation direction.
+	 * @returns {Boolean}
+	 */
+	_isSelectionAtCellEdge( cellRange, selection, isForward ) {
+		const model = this.editor.model;
+		const schema = this.editor.model.schema;
+
+		const focus = isForward ? selection.getLastPosition() : selection.getFirstPosition();
+
+		// If the current limit element is not table cell we are for sure not at the cell edge.
+		// Also `modifySelection` will not let us out of it.
+		if ( !schema.getLimitElement( focus ).is( 'tableCell' ) ) {
+			return false;
+		}
+
+		const probe = model.createSelection( focus );
+
+		model.modifySelection( probe, { direction: isForward ? 'forward' : 'backward' } );
+
+		// If there was no change in the focus position, then it's not possible to move the selection there.
+		return focus.isEqual( probe.focus );
+	}
+
+	/**
+	 * Checks if {@link module:engine/model/element~Element element} placed next to the current
+	 * {@link module:engine/model/selection~Selection model selection} exists and is marked in
+	 * {@link module:engine/model/schema~Schema schema} as `object`.
+	 *
+	 * @private
+	 * @param {module:engine/model/selection~Selection} modelSelection The selection.
+	 * @param {Boolean} forward Direction of checking.
+	 * @returns {module:engine/model/element~Element|null}
+	 */
+	_getObjectElementNextToSelection( modelSelection, forward ) {
```

Comment:
This method is only to detect cases when `Widget` plugin should handle this. It's trying to replicate behavior from original method: [`_getObjectElementNextToSelection`](https://github.com/ckeditor/ckeditor5-widget/blob/62735277c6f270b47a0c2723644e889c7b335dff/src/widget.js#L326).

Because of [stop propagation](https://github.com/ckeditor/ckeditor5-widget/blob/62735277c6f270b47a0c2723644e889c7b335dff/src/widget.js#L250) of keydown events, `TableNavigation` needs to listen on higher priority than `Widget`'s `keydown` listener.

Reply:
OK I didn't read the full history of this review. It is indeed copied from there.

Let's do a follow-up to extract it somewhere.